### DISCUSSION TITLE
Daylight Savings Time for USA

### DISCUSSION THREAD

**Discussion by Ken-Kauffmann** [arnaud-j4k08/PicoClockGreenEasy#3] [Q&A] [answered]
I am playing around with DST, and trying to modify DaylightSavingTime.cpp to include a section for USA in addition to Europe.
I have copied the block of code in isDstActive that calculates m_dstStart and m_dstEnd, and made modifications to conform to USA DST.

US DST starts on the 2nd Sunday in March, at 2:00am local time.  My changes are simple.  Modify the "dstStartTm.tm_hour = 1 + UTC_OFFSET" statement to read 2 instead of 1.   After the last Sunday in March is calculated, subtract either 14 or 21 from dstStartTm.tm_mday, depending upon whether the last Sunday occurred after the 28th ( 5 Sundays in March)

DST ends on the 1st Sunday in November, at 2:00 local time.  My changes were simple.  Change the dstEndTm.tm_hour calculation from 1 to 2 as above.  Modify the dstEndTm.tm_mon assignment from 9 to 10 to point to November. Modify the dstEndTm.tm_mday from 31 to 7 to point to November 7th.  Then use the dstEndTm.tm_wday subtraction to back us into the date for the first Sunday in November.

I made changes to UserConfig,cmake, and DaylightSavingTime.h to support USA in the enum of Location. 

I removed the battery from the RTC, and cleared the flash before loading the new uf2 file.

After making these changes, I tried to test this by changing the date and time to trigger the DST change.  They did not take effect.  I am not a C or C++ programmer, and I can easily have made a simple mistake, but it looks correct.   

Finally, a question (Actually 2 questions)

1)  What triggers the calculations in "isDstActive" (which modules call it)?
2)  Is DST inactivated by manually changing the Date and Time?

I am having fun.
Thanks, Ken

**Comment by arnaud-j4k08** [comment]
Hello Ken,

I'm glad you're having fun. Many people don't understand how cool it is to have a fully customizable alarm clock :-).

I'll have to look at this more carefully, as the US DST has some significant differences compared to the european one. In any case, you can test your changes by enabling tracing in this source code file. For this, enable tracing globally in UserConfig.cmake:
`set(TRACE_TO_STDIO "1")`
And enable tracing for this file by adding such a line in isEnabledForFile in Trace.cpp:
`file.find("DaylightSavingTime.cpp") != std::string::npos ||`
Then, you will be able to see such lines in a terminal monitoring the USB port on your PC (I use Tera Term for that):
```
.../DaylightSavingTime.cpp: 67: This year, DST starts on...
.../DaylightSavingTime.cpp: 79: This year, DST ends on...
```

You don't need to remove the RTC battery or clear the flash, as DST is considered permanently, so changes to the DST code take effect immediately.

To your questions:
> What triggers the calculations in "isDstActive" (which modules call it)?

isDstActive is a private method of the DaylightSavingTime class. It is notably used by DaylightSavingTime::considerDst, which is constantly called by Clock::setTmFromTime() to translate the current time without DST consideration to human readable time with DST consideration.

>Is DST inactivated by manually changing the Date and Time?

That's a tricky aspect which kept me busy for a while, so thanks for asking :-). Manually changing the Date and Time dynamically activates or deactivates DST depending on the timestamp within the year (outside or inside the DST period), so that the Date and Time always correspond to what the user expects to see. But additionally, within the hours of transition of the year (one when DST begins and one when DTS ends), being at this timestamp is ambiguous, as we wouldn't know if DST has already been applied or not. To handle that, the DaylightSavingTime::unconsiderDst method detects this ambiguity, and uses the previous result of isDstActive, stored in m_wasDstActive.

I hope this helps.

Best regards,
Arnaud

**Comment by Ken-Kauffmann** [comment]
It _is_ cool to have a fully customized alarm clock.  Especially when you can either sync it to UTC time, or pick some random Tuesday in 2026.

Thank you for the answers to the questions, and for the prompt to turn on tracing.  I had seen the instructions in UserConfig.cmake, but was a little scared to turn it on.  I also made a mistake the first time, and put the file.find statement in front of the return, instead of after.  Got more trace than I wanted, but quickly figured out how to get trace from DaylightSavingTime only.  I got some interesting results, and would like your input. 

Daylight Savings Time began in the US on March 10th, at 2:00am local time, which is 5 hours after UTC.  The result from the calculation was:
"This year, USA DST starts on 10/3/2024  wday: 0 21:0:0"
I wasn't sure if dstStartTm and dstEndTm were in UTC time, or Local time.  I am currently guessing that they are in Local time,    My UserConfig.cmake entry is "UTC_OFFSET=-5", which causes my NTP sync to give the correct local time.  I modified the following line to remove the + UTC_Offset " 
"dstStartTm.tm_hour = 2 + UTC_OFFSET;"
When I reran the test, and changed the year to 2022, and back to 2024, the result from the calculation was:
"This year, USA DST starts on 10/3/2024  wday: 0 2:0:0"

I changed the time to 1:55 am, and the date to March 3rd, the time correctly jumped to 3:00am when the time rolled over to 2:00.  I also started getting  "DST applied" traces.   Yay!

Similarly, Daylight Savings Time ended in the US on November 3rd, at 2:00am local time, which would be 4 hours after UTC. ( Offet of -5, with a skip forward of 1 hour due to DST).  The result from the initial calculation was:
"This year, USA DST ends on 3/11/2024  wday: 0 21:0:0"
Assuming that dstEndTm is also in local time, I modified the following line to remove the +UTC_OFFSET:
"dstEndTm.tm_hour = 2 + UTC_OFFSET; "
When I reran the text, and changed the year to 2025, and back to 2024, the result from the calculation was:
"This year, USA DST ends on 3/11/2024  wday: 0 2:0:0"

This looked promising.  I ran my test again, setting DST by changing the time to 1:55am on March 3rd, and then tried unsetting by 
changing the time to 1:55am on November 3rd.  I did not get the expected rollback when the time changed to 2:00am.  I got 2:00am
I did see that the "DST applied" traces ended, and I also noted that I got the trace message:

"In DST transition, use previous DST active flag"

...while I was modifying the date and time.    I am guessing that I need to find another way to test. 

So, after all that explanation, I have a few questions for you.

1) Am I correct in saying that dstStartTm and dstEndTm are in Local time, and not UTC?  

2) Do you agree that we do not have to apply the UTC_OFFSET to the hour assignment?  (If the offset is 0, as supplied in the original UserConfig.cmake, then the original statement works just fine)

3) Can you suggest a good way of testing the end of DST? (I have set the date to November the 2nd, 2024, at my current local time.  I will check in the morning to see if the change occurred overnight.  Heck, I might even stay up to 2:00.  I am currently getting "DST applied" traces. 

Thanks for your help.  Ken

**Reply by arnaud-j4k08** [reply]
> 1. Am I correct in saying that dstStartTm and dstEndTm are in Local time, and not UTC?

Yes, absolutely

>2. Do you agree that we do not have to apply the UTC_OFFSET to the hour assignment? (If the offset is 0, as supplied in the original UserConfig.cmake, then the original statement works just fine)

Yes, I agree. I use https://en.wikipedia.org/wiki/Daylight_saving_time_by_country as ultimate reference about daylight saving time in the world. Here you can see that, in contrast to European Union, US DST changes at a certain local time. So you need to remove all references to UTC_OFFSET from the block that calculates US DST start and end.

>3. Can you suggest a good way of testing the end of DST? (I have set the date to November the 2nd, 2024, at my current local time. I will check in the morning to see if the change occurred overnight. Heck, I might even stay up to 2:00. I am currently getting "DST applied" traces.

I just noticed that your problem is caused by a confusion that does not exist with the European DST: You need to set dstEndTm.tm_hour to the hour the DST ends, but expressed with DST not applied. So this line should be:
```
dstEndTm.tm_hour = 1;
```
Then, the trace will say:
```
"This year, USA DST ends on 3/11/2024 wday: 0 1:0:0"
```
which is, 02:00, right before DST stops to apply.

Also, when you manually change data and time, you may reach a timestamp that is ambiguous, because it actually exists "twice" from the user point of view. E.g. 2024-11-03 1:30 exists when DST still applies, and one hour later again, when DST no longer applies. That's what causes this message:
```
In DST transition, use previous DST active flag
```
Thus, after having made this correction, you can test the end of DST by setting your time to 1:58, your date to 2024-11-03, and wait two minutes. Make sure to set the time before setting the date, otherwise you may cross the DST transition hour, which will cause the ambiguity I just mentioned.

PS: In fact I hate DST. I find it useless and confusing, but I had to implement it.

**Comment by Ken-Kauffmann** [comment]
Success!  As you predicted, my test last night had the DST fallback occuring at 3:00am, rather than 2:00am.   The "DST applied"  traces stopped at that time, and all was right with the world.   This morning, I made the change to "dstEndTm.tm_hour = 1;"  that you suggested, and reran the tests as you described.  This time, fallback occurred at 2:00am on November 3rd, as expected.  I followed the steps as described, but still got 2 "In DST transition, use previous DST active flag" trace records during the process of changing from the activation of DST on March 3rd, and setting up the time for deactivation on November 3rd.  The "DST applied" traces still happened during the setup, so all was good.    I understand your "hatred" of DST.  There continues to be a lot of talk about eliminating it in the USA, and most people are behind that elimination.  The only reason for keeping it seems to be the number of programs and devices that would need to be updated.  The other issue is that half of the country wants to go back to Standard time year-round, and the other half wants to make DST permanent.  In any case, thank you for help in making this work.

**Comment by arnaud-j4k08** [comment, marked as answer]
I am glad it worked. The "In DST transition, use previous DST active flag" records are not a problem indeed. It's just something I used to test during the development.

>I understand your "hatred" of DST. There continues to be a lot of talk about eliminating it in the USA, and most people are behind that elimination. The only reason for keeping it seems to be the number of programs and devices that would need to be updated. The other issue is that half of the country wants to go back to Standard time year-round, and the other half wants to make DST permanent. In any case, thank you for help in making this work.

Aha, it seems to be same here in Europe. The European Parliament has basically agreed to discontinue DST for a while, but so far, the different countries failed to agree on which time to keep.

**Reply by arnaud-j4k08** [reply]
The last changes pushed to the repository add support for US DST. `DaylightSavingTime::Location` has a new `Usa` item that can be selected in UserConfig,cmake.
